src/analyses/apron/relationPriv.apron.ml

```diff
@@ -633,7 +643,9 @@ struct
           ) (RD.vars rel)
         in
         let rel_side = RD.keep_vars rel g_vars in
-        sideg V.mutex_inits rel_side;
+        (* If no globals are contained here, none need to be published *)
+        (* https://github.com/goblint/analyzer/pull/1354 *)
+        if g_vars <> [] then sideg V.mutex_inits rel_side;
```

Comment:
This sounds like some bottom-top confusion again: if there are no globals (i.e. the environment is empty), then don't side-effect (i.e. side-effect bottom) instead of side-effecting top.
Conditionally flipping a top to bottom seems inconsistent with all other values.

Reply:
I still think it's conceptually the right thing to do and not just a hack.
The thing is that we abuse the presence of globals in the relation as a no-overhead way to track whether one of the threads being joined possibly was in single-threaded mode.
If none of these threads was single-threaded (`g_vars = []`), all initial values are guaranteed to have been published and no side -effect is needed.

Does this make sense? If yes, I can write a comment detailing this reasoning.

Reply:
> The thing is that we abuse the presence of globals in the relation as a no-overhead way to track whether one of the threads being joined possibly was in single-threaded mode.

How so? We explicitly check the `MustBeSingleThreaded` query right before this. Prior to this, nothing is conditional on `g_vars = []`.

Reply:
This is the case of  branched thread creation. 

- **Case 0:** Both ST (single-threaded), nothing to do.
- **Case 1:** One thread is ST, and the other one is not. After the join, `Q.MustBeSingleThreaded` is false, and the globals from the single-threaded side of the join need need to be published. In this case, potentially, `g_vars <> []`. If `g_vars = []` there are no globals, and not causing a side-effect is ok.
- **Case 2:** Both MT, nothing should be done. In this case `g_vars = []`. Earlier $$\bot$$ was side-effected, not $$\top$$ would be side-effected which we need to protect against.

Reply:
Being multi-threaded doesn't imply `g_vars = []`, though.

Special-casing `[]` to flip the lattice is not a natural extension of the behavior on all other environments. I'm afraid we're setting ourselves up for a similar issue about top-bottom confusion that we're trying to fix here.

Suppose the global unknown has a non-bottom relation $R_{\\{x\\}}$ (on the environment $\\{x\\}$) and we're side-effecting $\top_\emptyset$ (on the empty environment). What is the join of these two relations?
1. $R_{\\{x\\}} \sqsubseteq R_{\\{x\\}}$ by reflexivity.
2. $\top_\emptyset \sqsubseteq R_{\\{x\\}}$ because $\emptyset \subseteq \\{x\\}$ and $\top_\emptyset \sqsubseteq R_{\\{x\\}}\vert_\emptyset = \top_\emptyset$ by the definition of heterogeneous $\sqsubseteq$ (e.g. see the unassume paper) and reflexivity.

It therefore follows that $R_{\\{x\\}}$ itself is an upper bound of $R_{\\{x\\}}$ and $\top_\emptyset$. Since it's equal to one of the arguments, it's obviously also the _least_ upper bound of the two.

So side-effecting $\top_\emptyset$ should be a no-op (and this special case is not needed) if we have the right lattice. If it isn't, then we still don't have the lattice fully right.

Reply:
Principled fix is delayed to #1380.